nomos-services/sdp/src/lib.rs

```diff
@@ -0,0 +1,119 @@
+}
+
+impl<B: SdpBackend<RuntimeServiceId> + 'static + Send + Sync, RuntimeServiceId> ServiceData
+    for SdpService<B, RuntimeServiceId>
+{
+    type Settings = B::Settings;
+    type State = NoState<Self::Settings>;
+    type StateOperator = NoOperator<Self::State>;
+    type Message = SdpMessage<B, RuntimeServiceId>;
+}
+
+pub struct SdpService<B: SdpBackend<RuntimeServiceId> + Send + Sync + 'static, RuntimeServiceId> {
+    backend: B,
+    service_state: OpaqueServiceStateHandle<Self, RuntimeServiceId>,
+}
+
+#[async_trait]
+impl<B, RuntimeServiceId> ServiceCore<RuntimeServiceId> for SdpService<B, RuntimeServiceId>
+where
+    B: SdpBackend<RuntimeServiceId> + Send + Sync + 'static,
+    RuntimeServiceId: AsServiceId<Self> + Clone + Display + Send,
+{
+    fn init(
+        service_state: OpaqueServiceStateHandle<Self, RuntimeServiceId>,
+        _initstate: Self::State,
+    ) -> Result<Self, overwatch::DynError> {
+        Ok(Self {
+            backend: B::new(
+                service_state.settings_reader.get_updated_settings(),
+                service_state.overwatch_handle.clone(),
+            ),
+            service_state,
+        })
+    }
+
+    async fn run(mut self) -> Result<(), overwatch::DynError> {
+        let Self {
+            mut backend,
+            service_state:
+                OpaqueServiceStateHandle::<Self, RuntimeServiceId> {
+                    mut inbound_relay,
+                    lifecycle_handle,
+                    ..
+                },
+        } = self;
+        let mut lifecycle_stream = lifecycle_handle.message_stream();
+        let backend = &mut backend;
+        loop {
+            tokio::select! {
+                Some(msg) = inbound_relay.recv() => {
+                    Self::handle_sdp_message(msg, backend).await;
+                }
+                Some(msg) = lifecycle_stream.next() => {
+                    if lifecycle::should_stop_service::<Self, RuntimeServiceId>(&msg) {
+                        break;
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+impl<B: SdpBackend<RuntimeServiceId> + Send + Sync + 'static, RuntimeServiceId>
+    SdpService<B, RuntimeServiceId>
+{
+    async fn handle_sdp_message(msg: SdpMessage<B, RuntimeServiceId>, backend: &mut B) {
+        match msg {
+            SdpMessage::Process {
+                block_number,
+                message,
+            } => {
+                if let Err(e) = backend.process_sdp_message(block_number, message).await {
+                    tracing::error!("Error processing SDP message: {}", e);
+                }
+            }
+            SdpMessage::MarkInBlock {
+                block_number,
+                result_sender,
+            } => {
+                let result = backend.mark_in_block(block_number).await;
+                // todo: should the result_sender been send to the backend?
+                let _ = result_sender.send(result);
```

Comment:
I think result sender should stay here, the only improvement would be to use tracing to track error just like in line `103` above.